File: r/tests/test_continuous_warning.R
#!/usr/bin/env Rscript
# Test continuous treatment variable warning

library(polars)
library(duckdb)
library(DBI)
source("R/common.R")
source("R/polars.R")
source("R/duckdb.R")

# Generate test data
set.seed(42)
n <- 1000

df <- pl$DataFrame(
  customer_id = rep(1:100, each = 10),
  product_id = rep(1:50, times = 20),
  treatment_continuous = runif(n, 0, 10),  # Continuous
  treatment_binary = rbinom(n, 1, 0.5)     # Binary
)

df <- df$with_columns(
  revenue = pl$lit(10.0) +
    pl$col("treatment_continuous") * 0.5 +
    pl$col("treatment_binary") * 1.0 +
    pl$col("customer_id")$cast(pl$Float64) * 0.01 +
    pl$lit(rnorm(n, 0, 1))
)

cat("Testing continuous treatment warnings...\n\n")

# Test 1: Continuous regressor triggers warning (Polars)
cat("Test 1: Continuous regressor warning (Polars)... ")
result <- tryCatch({
  suppressWarnings({
    leanfe_polars(df, formula = "revenue ~ treatment_continuous | customer_id + product_id")
  })
}, warning = function(w) {
  if (grepl("Continuous regressor", w$message)) {
    cat("WARNING DETECTED ✓\n")
    return(suppressWarnings(
      leanfe_polars(df, formula = "revenue ~ treatment_continuous | customer_id + product_id")
    ))
  }
})
stopifnot("treatment_continuous" %in% names(result$coefficients))
cat("PASSED\n")

# Test 2: Continuous regressor triggers warning (DuckDB)
cat("Test 2: Continuous regressor warning (DuckDB)... ")
result <- tryCatch({
  suppressWarnings({
    leanfe_duckdb(df, formula = "revenue ~ treatment_continuous | customer_id + product_id")
  })
}, warning = function(w) {
  if (grepl("Continuous regressor", w$message)) {
    cat("WARNING DETECTED ✓\n")
    return(suppressWarnings(
      leanfe_duckdb(df, formula = "revenue ~ treatment_continuous | customer_id + product_id")
    ))
  }
})
stopifnot("treatment_continuous" %in% names(result$coefficients))
cat("PASSED\n")

# Test 3: Binary treatment does NOT trigger warning
cat("Test 3: Binary treatment no warning... ")
warned <- FALSE
result <- tryCatch({
  leanfe_polars(df, formula = "revenue ~ treatment_binary | customer_id + product_id")
}, warning = function(w) {
  if (grepl("Continuous regressor", w$message)) {
    warned <<- TRUE
  }
  suppressWarnings(
    leanfe_polars(df, formula = "revenue ~ treatment_binary | customer_id + product_id")
  )
})
stopifnot(!warned)
stopifnot("treatment_binary" %in% names(result$coefficients))
cat("PASSED\n")

# Test 4: Mixed treatments trigger warning
cat("Test 4: Mixed treatments warning... ")
result <- tryCatch({
  suppressWarnings({
    leanfe_polars(df, formula = "revenue ~ treatment_binary + treatment_continuous | customer_id + product_id")
  })
}, warning = function(w) {
  if (grepl("Continuous regressor", w$message)) {
    cat("WARNING DETECTED ✓\n")
    return(suppressWarnings(
      leanfe_polars(df, formula = "revenue ~ treatment_binary + treatment_continuous | customer_id + product_id")
    ))
  }
})
stopifnot("treatment_binary" %in% names(result$coefficients))
stopifnot("treatment_continuous" %in% names(result$coefficients))
cat("PASSED\n")

# Test 5: Categorical treatment (few unique values) does NOT trigger warning
cat("Test 5: Categorical treatment no warning... ")
df_cat <- pl$DataFrame(
  customer_id = rep(1:100, each = 10),
  product_id = rep(1:50, times = 20),
  treatment_categorical = sample(c(0.0, 1.0, 2.0), n, replace = TRUE)
)
df_cat <- df_cat$with_columns(
  revenue = pl$lit(10.0) +
    pl$col("treatment_categorical") * 1.0 +
    pl$col("customer_id")$cast(pl$Float64) * 0.01 +
    pl$lit(rnorm(n, 0, 1))
)

warned <- FALSE
result <- tryCatch({
  leanfe_polars(df_cat, formula = "revenue ~ treatment_categorical | customer_id + product_id")
}, warning = function(w) {
  if (grepl("Continuous regressor", w$message)) {
    warned <<- TRUE
  }
  suppressWarnings(
    leanfe_polars(df_cat, formula = "revenue ~ treatment_categorical | customer_id + product_id")
  )
})
stopifnot(!warned)
stopifnot("treatment_categorical" %in% names(result$coefficients))
cat("PASSED\n")

cat("\n✅ All 5 tests passed!\n")

File: python/tests/test_continuous_warning.py
"""Test continuous treatment variable warning."""
import polars as pl
import numpy as np
import pytest
import warnings

from leanfe.polars_impl import leanfe_polars
from leanfe.duckdb_impl import leanfe_duckdb


@pytest.fixture
def continuous_treatment_data():
    """Generate test data with continuous treatment."""
    np.random.seed(42)
    n = 1000
    
    df = pl.DataFrame({
        'customer_id': np.repeat(np.arange(100), 10),
        'product_id': np.tile(np.arange(50), 20),
        'treatment_continuous': np.random.uniform(0, 10, n),  # Continuous
        'treatment_binary': np.random.binomial(1, 0.5, n),    # Binary
    })
    
    df = df.with_columns([
        (
            10.0 +
            0.5 * pl.col('treatment_continuous') +
            1.0 * pl.col('treatment_binary') +
            pl.col('customer_id').cast(pl.Float64) * 0.01 +
            pl.lit(np.random.normal(0, 1, n))
        ).alias('revenue')
    ])
    
    return df


def test_continuous_treatment_warning_polars(continuous_treatment_data):
    """Test that continuous treatment triggers warning in Polars."""
    with pytest.warns(UserWarning, match="Continuous regressor"):
        result = leanfe_polars(
            continuous_treatment_data,
            formula="revenue ~ treatment_continuous | customer_id + product_id"
        )
    
    # Should still run and produce results
    assert 'treatment_continuous' in result['coefficients']
    assert result['coefficients']['treatment_continuous'] > 0


def test_continuous_treatment_warning_duckdb(continuous_treatment_data):
    """Test that continuous treatment triggers warning in DuckDB."""
    with pytest.warns(UserWarning, match="Continuous regressor"):
        result = leanfe_duckdb(
            continuous_treatment_data,
            formula="revenue ~ treatment_continuous | customer_id + product_id"
        )
    
    # Should still run and produce results
    assert 'treatment_continuous' in result['coefficients']
    assert result['coefficients']['treatment_continuous'] > 0


def test_binary_treatment_no_warning_polars(continuous_treatment_data):
    """Test that binary treatment does NOT trigger warning."""
    with warnings.catch_warnings():
        warnings.simplefilter("error")  # Turn warnings into errors
        result = leanfe_polars(
            continuous_treatment_data,
            formula="revenue ~ treatment_binary | customer_id + product_id"
        )
    
    # Should run without warning
    assert 'treatment_binary' in result['coefficients']


def test_mixed_treatments_warning_polars(continuous_treatment_data):
    """Test that mixed treatments (binary + continuous) triggers warning."""
    with pytest.warns(UserWarning, match="Continuous regressor"):
        result = leanfe_polars(
            continuous_treatment_data,
            formula="revenue ~ treatment_binary + treatment_continuous | customer_id + product_id"
        )
    
    # Should still run and produce results for both
    assert 'treatment_binary' in result['coefficients']
    assert 'treatment_continuous' in result['coefficients']


def test_categorical_treatment_no_warning():
    """Test that categorical treatment (few unique values) does NOT trigger warning."""
    np.random.seed(42)
    n = 1000
    
    df = pl.DataFrame({
        'customer_id': np.repeat(np.arange(100), 10),
        'product_id': np.tile(np.arange(50), 20),
        'treatment_categorical': np.random.choice([0.0, 1.0, 2.0], n),  # Only 3 values
    })
    
    df = df.with_columns([
        (
            10.0 +
            1.0 * pl.col('treatment_categorical') +
            pl.col('customer_id').cast(pl.Float64) * 0.01 +
            pl.lit(np.random.normal(0, 1, n))
        ).alias('revenue')
    ])
    
    with warnings.catch_warnings():
        warnings.simplefilter("error")  # Turn warnings into errors
        result = leanfe_polars(
            df,
            formula="revenue ~ treatment_categorical | customer_id + product_id"
        )
    
    # Should run without warning (only 3 unique values)
    assert 'treatment_categorical' in result['coefficients']
